Content.Server/Kitchen/EntitySystems/ReagentGrinderSystem.cs

```diff
@@ -24,287 +25,252 @@ internal sealed class ReagentGrinderSystem : EntitySystem
+            var outputContainer = _itemSlotsSystem.GetItemOrNull(reagentGrinder.Owner, SharedReagentGrinder.BeakerSlotId);
+
+            // Do we have anything to grind/juice and a container to put the reagents in?
+            if (inputContainer.ContainedEntities.Count <= 0
+                || outputContainer is null
+                || !_solutionsSystem.TryGetFitsInDispenser(outputContainer.Value, out var containerSolution))
             {
                 return;
             }
 
-            switch (message.Message)
+            SoundSpecifier? sound = null;
+            switch (program)
             {
-                case SharedReagentGrinderComponent.ReagentGrinderGrindStartMessage msg:
-                    if (!this.IsPowered(component.Owner, EntityManager)) break;
-                    ClickSound(component);
-                    DoWork(component, attached,
-                        SharedReagentGrinderComponent.GrinderProgram.Grind);
+                case GrinderProgram.Grind when inputContainer.ContainedEntities.All(CanGrind):
+                    sound = reagentGrinder.GrindSound;
                     break;
-
-                case SharedReagentGrinderComponent.ReagentGrinderJuiceStartMessage msg:
-                    if (!this.IsPowered(component.Owner, EntityManager)) break;
-                    ClickSound(component);
-                    DoWork(component, attached,
-                        SharedReagentGrinderComponent.GrinderProgram.Juice);
+                case GrinderProgram.Juice when inputContainer.ContainedEntities.All(CanJuice):
+                    sound = reagentGrinder.JuiceSound;
                     break;
+                default:
+                    return;
+            }
 
-                case SharedReagentGrinderComponent.ReagentGrinderEjectChamberAllMessage msg:
-                    if (component.Chamber.ContainedEntities.Count > 0)
-                    {
-                        ClickSound(component);
-                        for (var i = component.Chamber.ContainedEntities.Count - 1; i >= 0; i--)
-                        {
-                            var entity = component.Chamber.ContainedEntities[i];
-                            component.Chamber.Remove(entity);
-                            entity.RandomOffset(0.4f);
-                        }
-
-                        EnqueueUiUpdate(component);
-                    }
-
-                    break;
+            _audioSystem.PlayPvs(sound, reagentGrinder.Owner);
+            reagentGrinder.Busy = true;
+            _userInterfaceSystem.TrySendUiMessage(reagentGrinder.Owner, ReagentGrinderUiKey.Key,
+                new ReagentGrinderWorkStartedMessage(program));
 
-                case SharedReagentGrinderComponent.ReagentGrinderEjectChamberContentMessage msg:
-                    if (component.Chamber.ContainedEntities.TryFirstOrNull(x => x == msg.EntityID, out var ent))
+            reagentGrinder.Owner.SpawnTimer(reagentGrinder.WorkTime, () => {
```

Comment:
it really shouldnt be using `SpawnTimer` (esp not the uid extensions), but if you dont want to change that here thats fine. just an accumulator and update loop would be preferable

Reply:
I added an `ActiveReagentGrinder` component with a timer. It counts down instead of up because checking if <0 is slightly faster than comparing to `reagentGrinder.WorkTime` every update. But if an accumulator is preferred I can change it.